feat(audit): more flatpak checks, adjust warning severity, JSON support

HastD opened the pull request:
Flatpaks are now checked for a list of dangerous filesystem access permissions, permissions that grant the ability to acquire arbitrary permissions, more device permissions, SSH agent permissions, and PulseAudio permissions. The flatpak permission checks also have a new `NOTICE` status used for some detected permissions, and the code has been separated out into its own `audit_flatpak` module and refactored to reduce code duplication.

Kernel argument checks have been separated into kargs automatically set by `ujust set-kargs-hardening` (which give a failure status if not present) and optional/unstable kargs (which only give a warning if not present), and alphabetized within each list.

The script now supports a `--json` (or `-j`) option to produce JSON output.

The formatting of the output has also been improved: the output is a little wider (up to 100 columns) when the terminal size allows for it, and ellipses rather than spaces are used for padding to make it easier to visually trace between a description and the corresponding status.

HastD committed:
feat(audit): more flatpak checks, adjust warning severity, JSON support

Flatpaks are now checked for "home" or "host" filesystem access,
permissions that grant the ability to acquire arbitrary permissions,
more device permissions, and PulseAudio permissions.

Kernel argument checks have been separated into kargs automatically set
by `ujust set-kargs-hardening` (which give a failure status if not
present) and optional/unstable kargs (which only give a warning if not
present).

A new status, `CAUTION`, has been added to allow for more fine-grained
feedback on flatpak permissions, and the warning level for various
flatpak permissions has been adjusting accordingly.

The script now supports a `--json` (or `-j`) option to produce JSON output.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
feat: clearer status names for flatpak checks; improve formatting

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

RoyalOughtness commented on files/system/usr/libexec/secureblue/auditor/__init__.py:
```diff
@@ -36,14 +37,44 @@ def to_str_in_color(self) -> str:
+    def to_str_in_color(self) -> str:
+        """Colored text representation of the status."""
+        match self:
+            case FlatpakStatus.SAFE:
```
what's the rationale for these? I don't think we should be projecting a sense of probability here. Our current method of using WARNING for potential sandbox escapes and FAILURE for known sandbox escapes makes more sense to me than this

HastD replied:
Okay, that's a helpful clarification, I was unsure about the rationale for what was marked as a "warning" vs "failure" for flatpaks.

The terms I used here are borrowed from [the Flathub website's app safety assessments](https://github.com/flathub-infra/website/blob/06706a4138d29225af9f8c49a73e9cb45991af59/frontend/src/safety.ts), but I agree that it'd be best not to overconfidently imply an app is "safe" just because its flatpak permissions pass a certain list of checks.

However, I'd still like to have at least four possible statuses for flatpaks rather than just "success, warning, failure". Roughly speaking, this is what I think makes sense for the different levels:

* Most severe level: known sandbox escape, broad system access (things like `filesystem=host` or `home`, anything that grants the ability to acquire arbitrary permissions).
* Middle level: potential sandbox escape, weakened security (e.g. not requesting `hardened_malloc`), access to some possibly sensitive systems.
* Less severe level: permissions like network access that are not likely sandbox escapes but are nonetheless worth informing the user about to reduce attack surface if not needed for the app's functionality.
* "Success" state: none of the above permissions detected.

Do you have suggestions for names for these? One possibility would be to use `FAILURE` and `WARNING` as before for the first two, and something like `NOTICE` for the third—what do you think of that? My main hesitation with this terminology is that certain apps—notably Flatseal itself—inherently require permissions that can act as sandbox escapes, meaning they'd always have a `FAILURE` status, so that terminology feels potentially misleading to the user.

Also, which permissions should go in which category? If the focus is on potential sandbox escapes then some of the prior categorizations were inconsistent; for example, network access was a "warning" despite not being a sandbox escape, and missing `hardened_malloc` was a "failure".

HastD replied:
Anyway for now I changed the status labels back, with an added `NOTICE` label like I described above. I still think it's worth thinking through more how different permissions are categorized and maybe adjusting some of them, but no need to make that decision in this PR, I suppose.

RoyalOughtness replied:
@HastD what about

success -> SUCCESS
less severe -> INFO
middle -> WARNING
most severe -> FAILURE

HastD replied:
Would you prefer `INFO` to `NOTICE`? I don't have a strong opinion between the two.

HastD replied:
...actually on second thought I think do I prefer "notice"; it feels like it has more of a connotation of "this is something you might want to act on", which is appropriate since those are still permissions that shouldn't be given out unnecessarily, even if they have less serious security implications.

RoyalOughtness replied:
@HastD agreed, NOTICE makes sense

HastD committed:
Use "success/notice/warning/failure" status labels for flatpak checks.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
improve visuals of line between report description and status

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -800,45 +806,99 @@ async def check_flatpak_permissions(name, version, state):
         sockets = perms["sockets"]
         if "x11" in sockets and "fallback-x11" not in sockets:
-            status = FAILURE
+            status = status.downgrade_to(WARNING)
```
x11 -> failure

?

HastD replied:
I guess it does count as a sandbox escape, if only between x11 apps and not to the whole system... so sure, I could change it back to failure.

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -780,16 +787,15 @@ async def check_flatpak_permissions(name, version, state):
         if "network" in shared:
-            if status != FAILURE:
-                status = WARNING
+            status = status.downgrade_to(NOTICE)
```
yeah NOTICE or INFO makes sense to me

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -800,45 +806,99 @@ async def check_flatpak_permissions(name, version, state):
+        devices = perms["devices"]
+        device_checks = {
+            "all": {
+                "status": WARNING,
```
FAILURE?

HastD replied:
Since it can be used as a sandbox escape, sure. For the same reason I'll also change `device=shm` to `FAILURE`.

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -859,35 +919,72 @@ async def check_flatpak_permissions(name, version, state):
+                    $ flatpak override -u --nofilesystem=host {name}"""
+            )
+        if "home" in filesystems:
+            status = FAILURE
```
downgrade_to? I guess it's not needed... but it would be consistent

HastD committed:
add device=usb check, adjust a few status levels for flatpak checks

RKNF404 commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -800,45 +806,105 @@ async def check_flatpak_permissions(name, version, state):
+                "note": "",
+            },
+            "usb": {
+                "status": NOTICE,
```
This should be a warning, depending on what USB devices are plugged in, it may be a sandbox escape as well

HastD replied:
Oh right, like if there's a USB keyboard or something I can see how that could be a sandbox escape...

RKNF404 commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -859,35 +925,72 @@ async def check_flatpak_permissions(name, version, state):
+                    $ flatpak override -u --nofilesystem=home {name}"""
+            )
+
+        prefixes = ("xdg-data", "~/.local/share")
```
This check should probably be expanded a bit to other accesses, such as xdg-run which allows the app to gain access to any socket even ones flatpak doesn't allow

HastD replied:
I agree that there should be a check for `xdg-run` access. However, this specific check is for permissions that give the ability to acquire arbitrary permissions by modifying flatpak overrides. Is access to `xdg-run` in that category, or should it be part of a separate check for dangerous filesystem permissions?

RKNF404 replied:
likely separate, just pointing out that xdg-data isnt the only risky access we should be checking for here

RKNF404 commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -800,45 +806,105 @@ async def check_flatpak_permissions(name, version, state):
         if "system-bus" in sockets:
-            if status != FAILURE:
-                status = WARNING
+            status = status.downgrade_to(WARNING)
```
Session and system bus should be failures, no? They are literally sandbox escape permissions

HastD committed:
feat(audit): add more flatpak permission checks, adjust others

* Check for filesystem=xdg-cache, xdg-config, xdg-data permissions.
* Improve filesystem permissions parsing.
* Make session-bus, system-bus permissions a FAILURE.
* Make device=usb permissions a WARNING.

Also, a few minor style/formatting changes:
* Alphabetize karg checks for readability.
* Make punctuation consistent in recommendations.
* Remove extra '...' from reports.

HastD committed:
refactor(audit): move flatpak permission checks to separate module

HastD committed:
refactor(audit): reduce code duplication in flatpak permission checks

The new class PermissionCheck encodes the necessary data for many basic
permission checks, allowing a lot of repetitive code to be factored out.

Also, parsing of flatpak permissions has been made more robust.

HastD committed:
feat(audit): add flatpak permission check for socket=ssh-auth

RoyalOughtness committed:
Merge branch 'live' into audit-flatpak-permissions
